core/src/cluster_slot_state_verifier.rs

```diff
@@ -2,24 +2,36 @@ use crate::{
     fork_choice::ForkChoice, heaviest_subtree_fork_choice::HeaviestSubtreeForkChoice,
     progress_map::ProgressMap,
 };
+use solana_ledger::blockstore::Blockstore;
 use solana_sdk::{clock::Slot, hash::Hash};
 use std::collections::{BTreeMap, BTreeSet, HashSet};
 
 pub(crate) type DuplicateSlotsTracker = BTreeSet<Slot>;
 pub(crate) type DuplicateSlotsToRepair = HashSet<(Slot, Hash)>;
 pub(crate) type GossipDuplicateConfirmedSlots = BTreeMap<Slot, Hash>;
-type SlotStateHandler = fn(Slot, &Hash, Option<&Hash>, bool, bool) -> Vec<ResultingStateChange>;
+type SlotStateHandler = fn(SlotStateHandlerArgs) -> Vec<ResultingStateChange>;
+
+#[derive(Debug)]
+struct SlotStateHandlerArgs {
+    slot: Slot,
+    bank_frozen_hash: Hash,
+    cluster_duplicate_confirmed_hash: Option<Hash>,
+    is_dead: bool,
+    is_slot_duplicate: bool,
+}
 
 #[derive(PartialEq, Debug)]
 pub enum SlotStateUpdate {
-    Frozen,
+    BankFrozen,
     DuplicateConfirmed,
     Dead,
     Duplicate,
 }
```

Comment:
I think it'd be awesome to add some extra type safety and remove the need for the assertions in this module. With these adjustments you could also remove some test cases that are nonsensical.

```rust
pub enum BankStatus {
  Frozen(Hash),
  Dead,
  Unprocessed,
}

pub enum SlotStateUpdate {
    BankFrozen(Hash),
    DuplicateConfirmed(BankStatus),
    Dead,
    Duplicate(BankStatus),
}
```

Reply:
Updated!